### PR TITLE
test(api): prove released MCP reconciliation and post-deploy convergence

### PR DESCRIPTION
**Pull request by yomgui**
## Summary
- Black-box curl/jq on pinned Den API 0.18.46 (linux/arm64): 8/8 MCP cases, including same-image tenant isolation and verbatim HTTPS recipe.
- Live dev testkit covers eight cases. Corrected full post-deploy job executed twice: seven convergent mutations all 200 on run 2, identical counts/IDs. Intentional marketplace deletion/recreation is separate.
- Provider usability, unchanged-identity OAuth client-secret retention, and unkeyed MCP ID updates verified. Omitted MCP access widens to org-wide; released OpenAPI version says dev; members/teams lists come from /v1/org.
- Historical signed and signed-off commits; no product runtime change, merge, or production deployment. Sanitized full receipts and reproduction fixtures included.

## Proof and limitations
Consolidated report: reports/mcp-put-by-key-proof-2026-09-14.md
Executable: examples/declarative-org/post-deploy-job.sh
Historical local narrow runs before this cleanup: 10 release-receipt, 10 job-receipt, 6 redaction, 1 live-dev tests passed; zero failures/skips. The historical external evidence link has been removed; see the confidentiality boundary below.

**Overall integration verdict: Incomplete — closed, not for merge.** The boundary ratchet rejects the three new offline receipt/redaction specs for lacking a live product boundary. They are explicitly historical receipt checks, not released-image reexecution. They need approved placement or genuine released-image test-boundary integration before review-ready; no policy bypass/baseline expansion added. Full typecheck also has two unresolved diagnostics in untouched files; full layer lint has 54 out-of-scope violations. No clean-control claim that these are pre-existing.

This is scoped runtime proof, not a blanket guarantee of all OAuth providers, concurrent first applies, amd64, or unobserved deployments. Setup used explicit headless admin bootstrap APIs; actual provisioning requests used API keys only. Private credentials/logs were not committed.

## Confidentiality cleanup
Retired witness links in tracked receipts were replaced with reserved example domains; the job filename and recorded path are neutralized. Integrity checks now refer to the sanitized copies, not the original historical bytes. Explicit synthetic fixtures are allowlisted in the consolidated report. The external review-report/image link was removed from the author-owned evidence comment.

This does not erase prior Git objects, hosted media, caches, or earlier comment revisions. Bot-authored findings and deployment metadata are not edited. No product fix, new release execution, merge, or rehoming is claimed. This PR must remain closed.

The retained branch was fast-forward updated to `9fd50c31b3e08de4581db3ad6efd9be1c998ceee`. GitHub still exposes this closed PR’s head ref and Files view at the prior commit; those historical copies remain accessible. The cleanup statements above describe the retained branch, not erasure of the frozen PR diff. This limitation is not addressed by reopening or rewriting history.

Follow-up on the retained branch replaces all opaque recorded runtime identities with stable neutral role/ordinal aliases and uses `demo-proof-*` fixture keys. Equality relationships and outcomes are preserved; dependent checksums and sanitized-copy notices are updated. Actual opaque runtime identities are no longer allowlisted. The closed-PR historical-ref limitation above still applies.

### CODE REVIEW

**Comment by yomgui** [comment]
Historical verification summary: 12/12 selected tests and 18/18 assertions were reported for commit 3abfa41b4240d9968cd471f3057e3deab83c43ed.

The external review report and its eight images are no longer linked here for confidentiality. Removing this link does not erase hosted media, caches, or prior revisions. This is not fresh verification of the current branch. The PR remains closed and is not for merge.

File: evals/fixtures/mcp-put-release-witness.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        if MODE == 'proxy':
+            connection = http.client.HTTPConnection('127.0.0.1', 18788, timeout=90)
+            headers = {k: v for k, v in self.headers.items() if k.lower() not in ['host', 'connection']}
+            connection.request(self.command, self.path, body=raw, headers=headers)
+            upstream = connection.getresponse()
+            result = upstream.read()
+            self.send_response(upstream.status)
+            for k, v in upstream.getheaders():
+                if k.lower() not in ['transfer-encoding', 'content-length', 'connection']:
+                    self.send_header(k, v)
```
[the remark this answers is not in the document]
Comment (yomgui) [reply evals/fixtures/mcp-put-release-witness.py]: Disposition: TRUE POSITIVE unsafe forwarding boundary (header value), fixed in 3abfa41b4240d9968cd471f3057e3deab83c43ed. In particular, folded upstream field values can retain CR/LF when forwarded.

Options considered:
- Strip CR/LF alone: prevents line splitting but leaves invalid control characters.
- Drop all upstream headers or use a fixed allowlist: changes the proxy proof semantics unnecessarily.
- Chosen: validate all retained upstream names/values before any response is emitted, return 502 and close upstream on invalid input, and strip CR/LF at the final send_header boundary. The same fix addresses the separate name finding 322.

Verification: field-value cases cover CR, LF, CRLF, NUL, DEL, and non-Latin-1 input; no injected header is emitted. Valid tab-containing values and content type survive, status/body remain intact, and Content-Length is recomputed. Python unittest and py_compile passed after this shared fix; 34 exact-head testkit tests pass, zero skips. Python CodeQL analysis 1789212697 has zero findings and alert 323 reports fixed.
Evidence and limits: https://github.com/different-ai/openwork/pull/4983#issuecomment-5667623292

File: evals/fixtures/mcp-post-deploy-release.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    inspection['composeSha256'] = hashlib.sha256(pathlib.Path('packaging/docker/docker-compose.eval.yml').read_bytes()).hexdigest()
+    witness = [json.loads(line) for line in (ROOT / 'witness.jsonl').read_text().splitlines()]
+    output = {'schemaVersion': 2, 'kind': 'recorded-release-post-deploy-job', 'recordingRevision': 2, 'recordedAt': datetime.datetime.now(datetime.timezone.utc).isoformat(), 'supersedesTranscriptSha256': supersedes, 'inspection': inspection, 'jobSha256': execution_sha, 'jobInvocations': clean(applies), 'otherInvocations': clean([row for row in invocations if row['mode'] != 'apply']), 'priorTranscripts': immutable, 'comparison': json.loads((ROOT / 'comparison.json').read_text()), 'credentialChecksBeforeRedaction': checks, 'expectedSyntheticClientSecretFingerprint': hashlib.sha256(state['oauthSecret'].encode()).hexdigest()[:16], 'expectedSyntheticProviderSecretFingerprint': hashlib.sha256(state['providerSecret'].encode()).hexdigest()[:16], 'testkit': 'deferred_to_orchestrator', 'witnessReceipts': clean(witness), 'requests': clean(receipts)}
```
[the remark this answers is not in the document]
Comment (yomgui) [reply evals/fixtures/mcp-post-deploy-release.py]: Disposition: FALSE POSITIVE for password hashing, with fixture hardening in 3abfa41b4240d9968cd471f3057e3deab83c43ed. These values are random 32-byte synthetic credentials, and the truncated SHA256 was an equality fingerprint, not a password verifier.

Options considered:
- Keep plain SHA256 and explain the high-entropy fingerprint context: valid, but retains the ambiguous password-hash warning.
- Use a password KDF or remove fingerprints: unnecessary for random tokens, or loses the cross-witness retention assertion.
- Chosen: deterministic HMAC-SHA256 keyed by the synthetic credential over a fixed, versioned proof domain, in both exporter and witness. No new persisted key or suppression; file-integrity SHA256 and historical transcript bytes remain unchanged.

Verification: matching exporter/witness fingerprints, rotation changes, redacted token logs, and the OAuth/PKCE positive and negative controls pass. Python unittest and py_compile ran after the change. Final exact-head testkit runs: 34 passed, zero skips. Python CodeQL analysis 1789212697 has zero findings and alert 320 reports fixed.
Evidence and limits: https://github.com/different-ai/openwork/pull/4983#issuecomment-5667623292

File: evals/fixtures/mcp-put-release-witness.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    do_POST = handle_request
+    do_PUT = handle_request
+    do_DELETE = handle_request
+
+
+server = http.server.ThreadingHTTPServer(('127.0.0.1' if MODE == 'proxy' else '0.0.0.0', 18443 if MODE == 'proxy' else 8080), Handler)
+if MODE == 'proxy':
+    context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+    context.load_cert_chain(os.environ['PROOF_TLS_CERT'], os.environ['PROOF_TLS_KEY'])
+    server.socket = context.wrap_socket(server.socket, server_side=True)
```
[the remark this answers is not in the document]
Comment (yomgui) [reply evals/fixtures/mcp-put-release-witness.py:81]: Disposition: TRUE POSITIVE configuration gap, fixed in 3abfa41b4240d9968cd471f3057e3deab83c43ed. The fixture depended on runtime TLS defaults rather than declaring its minimum version; this does not claim that modern Python 3.12 defaults accepted TLS 1.0.

Options considered:
- Rely on current SSLContext defaults: leaves the floor runtime-dependent.
- Require TLS 1.3 only: unnecessarily excludes valid TLS 1.2 proof clients.
- Chosen: retain PROTOCOL_TLS_SERVER and explicitly set minimum_version = TLSv1_2 before certificate loading and server-side wrapping. No suppression or product change.

Verification: startup is exercised with a real SSLContext initially set to MINIMUM_SUPPORTED; the fixture raises it to TLS 1.2 and wraps server-side with the configured certificate paths. The test checks the configured policy, not a live network handshake. Python unittest and py_compile passed after the fix; 34 exact-head testkit tests pass, zero skips. Python CodeQL analysis 1789212697 has zero findings and alert 319 reports fixed.
Evidence and limits: https://github.com/different-ai/openwork/pull/4983#issuecomment-5667623292

File: evals/fixtures/mcp-put-release-witness.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yomgui) [reply evals/fixtures/mcp-put-release-witness.py]: Disposition: TRUE POSITIVE unsafe forwarding boundary (header name), fixed in 3abfa41b4240d9968cd471f3057e3deab83c43ed. The sink forwarded upstream header names without a local token check.

Options considered:
- Trust the upstream parser or suppress the finding: retains an unchecked send_header boundary.
- Forward only a fixed header allowlist: unnecessarily reduces the proxy evidence and behavior.
- Chosen: collect the retained upstream headers, validate every name as an HTTP token and every value before sending a status/header, reject invalid input with 502 and close upstream, then strip CR/LF from names and values at send_header.

Verification: header-name cases cover CR, LF, CRLF, colon, space, NUL, and non-Latin-1 input; no injected header is emitted. Valid response headers/status/body are preserved, hop-by-hop/framing headers remain excluded, and Content-Length is recomputed. Python unittest and py_compile passed after the shared 322/323 fix; 34 exact-head testkit tests pass, zero skips. Python CodeQL analysis 1789212697 has zero findings and alert 322 reports fixed.
Evidence and limits: https://github.com/different-ai/openwork/pull/4983#issuecomment-5667623292

File: evals/fixtures/mcp-put-release-client.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    secret_checks = [{'label': r['label'], 'rawResponseContainsWitnessSecret': secret in json.dumps(r['response']), 'responseHasApiKeyField': 'apiKey' in r['response']['body']} for r in receipts if r['label'] in ['case6-secret-create', 'case6-secret-read', 'case6-omit-secret', 'case6-secret-read-after']]
+    witness = [json.loads(line) for line in (ROOT / 'witness.jsonl').read_text().splitlines()]
+    output = {'schemaVersion': 1, 'kind': 'recorded-release-blackbox', 'image': 'ghcr.io/different-ai/openwork-den-api:0.18.46@sha256:e254a3842e7ecb6d0c7128b5f17201e92ac4ee566ad82c70c468938a3faa6407', 'inspection': inspection, 'coverage': {'foreignOrgId': 'not_executed_single_org_bootstrap_409', 'testkit': 'deferred_to_orchestrator'}, 'secretChecksBeforeRedaction': secret_checks, 'witnessCalls': [row for row in witness if row['method'] == 'tools/call'], 'requests': clean(receipts)}
+    serialized = json.dumps(output, indent=2) + '\n'
+    if any(value and value in serialized for value in secrets):
+        raise RuntimeError('Secret remains in transcript; refusing export')
+    pathlib.Path(destination).write_text(serialized)
```
[the remark this answers is not in the document]
Comment (yomgui) [reply evals/fixtures/mcp-put-release-client.py:109]: Disposition: FALSE POSITIVE at this reported source/sink, plus export hardening in 3abfa41b4240d9968cd471f3057e3deab83c43ed. The flagged source is the entire list assigned to secret_checks; it contains labels and boolean membership checks, not credential bytes. CodeQL SensitiveVariableAssignment uses sensitive-name heuristics.

Options considered:
- Suppress the query: rejected; no suppression is needed.
- Remove or encrypt the boolean evidence: loses readable retention checks without protecting an actual secret.
- Chosen: accurately name the metadata response_checks and sanitize the entire exported object, including previously uncleaned witnessCalls. Raw comparison inputs remain private; exported metadata is boolean and the existing final secret-leak guard remains.

Verification: the actual export function processes synthetic true/false checks and a witness nonce containing a generated credential; persisted output retains boolean checks, redacts that nonce and unknown auth/cookie values, and contains no credential. Python unittest and py_compile passed after this change; 34 exact-head testkit tests pass, zero skips. Python CodeQL analysis 1789212697 has zero findings and alert 318 reports fixed.
Evidence and limits: https://github.com/different-ai/openwork/pull/4983#issuecomment-5667623292

File: evals/fixtures/mcp-post-deploy-witness.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                    payload = {}
+                result = {'jsonrpc': '2.0', 'id': body['id'], 'result': payload}
+        else:
+            status, result = 404, {'error': 'not_found'}
+        encoded = json.dumps(result).encode()
+        self.send_response(status)
+        self.send_header('Content-Type', 'application/json')
+        self.send_header('Content-Length', str(len(encoded)))
+        for key, value in headers.items():
+            self.send_header(key, value)
```
[the remark this answers is not in the document]
Comment (yomgui) [reply evals/fixtures/mcp-post-deploy-witness.py]: Disposition: TRUE POSITIVE, fixed in 3abfa41b4240d9968cd471f3057e3deab83c43ed. Decoded redirect_uri and folded Host input can reach Location or WWW-Authenticate values.

Options considered:
- Strip CR/LF only: prevents splitting but tolerates other invalid header controls.
- Validate only: sufficient rejection, but lacks the explicit final CR/LF sanitizer.
- Chosen: validate header token names and permitted field-value characters before sending any response headers; reject invalid values with 400, and strip CR/LF from both names and values at send_header as defense in depth.

Verification: real handler serialization rejects CR, LF, CRLF, NUL, DEL, non-Latin-1 redirect inputs and folded Host input without emitting injected headers. Valid OAuth redirect, PKCE exchange, authenticated MCP/provider calls, wrong-secret/verifier controls, and code replay checks pass. Python unittest and py_compile passed after the fix; all 34 exact-head testkit tests pass. Python CodeQL analysis 1789212697 has zero findings and alert 321 reports fixed.
Evidence and limits: https://github.com/different-ai/openwork/pull/4983#issuecomment-5667623292

**Comment by yomgui** [comment]
Closing without merge: this proof was produced for a specific customer engagement and is kept as a reference only; engagement-specific material does not land in the public repository. The generic API behaviours it exercised are covered by the merged declarative-configuration example and its spec. Branch retained.
